analyzer/protocol/openvpn/openvpn.spicy

```diff
@@ -0,0 +1,181 @@
+		Opcode::P_ACK_V1 -> ack_v1: AckMessage(self.message_type.opcode, self.message_type.key_id, False, share) &size=self.packet_length-1;
+		Opcode::P_DATA_V1 -> data_v1: DataMessage(self.message_type.opcode, self.message_type.key_id, False, share) &size=self.packet_length-1;
+		Opcode::P_CONTROL_HARD_RESET_SERVER_V2 -> control_hard_reset_server_v2: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share) &size=self.packet_length-1;
+		Opcode::P_CONTROL_HARD_RESET_CLIENT_V2 -> control_hard_reset_client_v2: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share) &size=self.packet_length-1;
+		Opcode::P_DATA_V2 -> data_v2: DataMessage(self.message_type.opcode, self.message_type.key_id, True, share) &size=self.packet_length-1;
+	};
+};
+
+public type OpenVPNRecordsTCPHMAC = unit {
+	%context=Share;
+	records : OpenVPNRecordTCPHMAC(self.context())[];
+};
+
+public type OpenVPNRecordTCPHMAC = unit(inout share: Share) {
+	packet_length : uint16;
+
+	message_type: bitfield(8) {
+		opcode: 3..7 &convert=Opcode($$);
+		key_id: 0..2;
+	};
+
+	switch ( self.message_type.opcode ) {
+		Opcode::P_CONTROL_HARD_RESET_CLIENT_V1 -> control_hard_reset_client_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share) &size=self.packet_length-1;
+		Opcode::P_CONTROL_HARD_RESET_SERVER_V1 -> control_hard_reset_server_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share) &size=self.packet_length-1;
+		Opcode::P_CONTROL_SOFT_RESET_V1 -> control_soft_reset_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share) &size=self.packet_length-1;
+		Opcode::P_CONTROL_V1 -> control_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share) &size=self.packet_length-1;
+		Opcode::P_ACK_V1 -> ack_v1: AckMessage(self.message_type.opcode, self.message_type.key_id, True, share) &size=self.packet_length-1;
+		Opcode::P_DATA_V1 -> data_v1: DataMessage(self.message_type.opcode, self.message_type.key_id, False, share) &size=self.packet_length-1;
+		Opcode::P_CONTROL_HARD_RESET_SERVER_V2 -> control_hard_reset_server_v2: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share) &size=self.packet_length-1;
+		Opcode::P_CONTROL_HARD_RESET_CLIENT_V2 -> control_hard_reset_client_v2: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share) &size=self.packet_length-1;
+		Opcode::P_DATA_V2 -> data_v2: DataMessage(self.message_type.opcode, self.message_type.key_id, True, share) &size=self.packet_length-1;
+	};
+};
+
+type HMACInfo = unit {
+	hmac : bytes &size=20;
+	packet_id : uint32;
+	net_time  : bytes &size=4;
+};
+
+public type ControlMessage = unit(opcode: Opcode, key_id: uint8, has_hmac: bool, inout share: Share) {
+	var opcode: Opcode = opcode;
+	var key_id: uint8 = key_id;
+	session_id : bytes &size=8;
+	switch ( has_hmac ) {
+		True -> hmac: HMACInfo;
+		False -> no_hmac: bytes &size=0;
```

Comment:
You can use `void` as the type here:

```
False -> no_hmac: void;
```

Or a conditional field instead of the switch:

```
hmac: HMACInfo if ( self.has_hmac )
````

Reply:
Nice!